components/relevancy/src/lib.rs

```diff
@@ -0,0 +1,80 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Proposed API for the relevancy component (validation phase)
+//!
+//! The goal here is to allow us to validate that we can reliably detect user interests from
+//! history data, without spending too much time building the API out.  There's some hand-waving
+//! towards how we would use this data to rank search results, but we don't need to come to a final
+//! decision on that yet.
+
+/// List of possible interests for a domain.  Each domain would have exactly one of these.  If we
+/// can't calculate an interest based on the classifier data, then we put it in `Unknown`
+#[derive(uniffi::Enum)]
+pub enum Interest {
+    Education,
+    Gaming,
+    Sports,
+    Unknown,
+}
+
+/// An interest combined with a relevancy score
+///
+/// Scores will be [0-10].  Consumers would use this score to alter the order of their suggestions.
+///
+/// For example, consumers may assign a default score of 5 to sponsored suggestions and 0 to
+/// Wikipedia suggestions.  By default, the sponsored suggestions will show up over Wikipedia
+/// suggestions, but if a Wikipedia suggestion had a interest that was highly relevant for the
+/// user, that suggestion would show up first instead.
+#[derive(uniffi::Record)]
+pub struct InterestScore {
+    interest: Interest,
+    score: u8,
+}
+
+/// Top-level class that ingests and stores relevancy data.
+///
+/// For the validation phase:
+///   - All data will be stored in memory.
+///   - Consumers will call `ingest` then `get_top_categories` and store the results in a Glean
+///     distribution (see `metrics.yaml).  Our main focus will be to monitor these metrics in order
+///     to determine if the relevancy scores would be strong enough to affect search rankings.
+///   - If possible, I'd also like to have make "about::relevancy" where users can go to see the
+///     results themselves.  This way we can pair the Glean metrics with asking people if they
+///     think the interest rankings seem correct to them.
+#[derive(uniffi::Object)]
+pub struct RelavancyStore {
+}
+
+#[uniffi::export]
+impl RelavancyStore {
+    /// Ingest relevancy data from remote settings
```

Comment:
Here's my proposal for storing the data in remote settings:

- The data is stored in a binary attachment
- In that file, for each interest we store:
   - The count of URLs for that interest (32 bits)
   - For each URL, we store a SHA-256 hash of that URL
- In the Rust component, we build a `HashSet<u8[32], Topic>` to map URL hashes to topics.  (Assuming URLs, can only have 1 topic.  If not, then we can use `Vec<Topic>` as the value).
  
This is similar to the bloom filter approach, but the hashing is simpler.  If we use a bloom filter, then I think it would be somewhat tricky to get the server and the component to agree on the hashing system.  SHA-256 avoids those issues.

I don't know how to think about the size differences, but I believe it should be similar.  Lookups will be slower, but I think a `HashSet` should perform well enough for our needs.

Reply:
(Maybe MD5 would be be better, since it outputs 128 bit value, which we could store in a `u128`).  The collision chance is still very low, and if we did hit a collision that wouldn't be the end of the world -- probably it's a negligible loss in accuracy).

Reply:
I think that's a solid proposal for now, love it!

The current dataset, each domain could have more than 1 category, but we can simplify that by only assigning one category (the good news is we also have a score for each category, so we could choose the one with the higher score).

Using MD5 is totally fine, actually we encode URLs via MD5 extensively in Desktop. This is quite similar to how we store domains/categories for the SERP search term categorization project. Expect a GH invite to that repo so that you can see more details there.

Lastly, I think we might want to add another field on RS indicating what part of URLs used (host + TLD or subdomain + host + TLD) so that when we classify interests, we can extract the right parts as well.